File: programmers/베스트앨범.py
"""
코딩테스트 연습 > 해시
"""


def solution(genres, plays):
    genres_map = {}
    for i in range(len(genres)):
        d = (i, plays[i])
        if genres[i] in genres_map.keys():
            genres_map[genres[i]].append(d)
        else:
            genres_map[genres[i]] = [d]

    data = []
    for musics in genres_map.values():
        genres_play = 0
        for m in musics:
            genres_play += m[1]

        musics.sort(key=lambda x: -x[1])

        if len(musics) > 2:
            musics = musics[:2]
        for m in musics:
            data.append((m[0], genres_play, m[1]))

    data.sort(key=lambda x: (-x[1], -x[2], x[0]))

    return list(map(lambda x: x[0], data))

print(solution(["classic", "pop", "classic", "classic", "pop"], [500, 600, 150, 800, 2500]))

File: programmers/베스트앨범.kt
class Solution {
    fun solution(genres: Array<String>, plays: IntArray): IntArray {
        var genresMap = (0..genres.size-1).map {
            genres[it] to plays[it]
        }.groupBy {
            it.first
        }.mapValues { (k, v) ->
            v.sumOf { it.second }
        }

        var topGenres = genresMap
            .toList()
            .sortedByDescending{
                    (k,v) -> v
            }.map {
                it.first
            }

        var playInGenresMap = (0..genres.size-1).map {
            listOf(genres[it], it, plays[it])
        }.groupBy {
            it[0]
        }.mapValues { (k, v) ->
            v.sortedWith(
                compareBy(
                    { -(it[2] as Int) },
                    { (it[1] as Int) }
                )
            )
        }

        var result = mutableListOf<Int>()
        topGenres.forEach {
            var songInGenres = playInGenresMap[it]!!
            result.add(songInGenres[0][1] as Int)
            if (songInGenres.size > 1) {
                result.add(songInGenres[1][1] as Int)
            }
        }

        return result.toIntArray()
    }
}
